cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -418,6 +418,9 @@ AzureOptions::MakeDataLakeServiceClient() const {
 
 Result<std::string> AzureOptions::GenerateSASToken(
     Storage::Sas::BlobSasBuilder* builder, Blobs::BlobServiceClient* client) const {
+  using SasProtocol = Storage::Sas::SasProtocol;
+  builder->Protocol =
+      blob_storage_scheme == "http" ? SasProtocol::HttpsAndHttp : SasProtocol::HttpsOnly;
```

Comment:
Can we validate the value is either "http" or "https"? We shouldn't allow other values silently.

Reply:
We can validate the value explicitly but other operations such as creating `BlobServiceClient` or `DataLakeServiceClient` will be failed with invalid value before this code.
Should we validate the value explicitly? If we should validate the value, we'll do it in `MakeBlobServiceClient()` and `MakeDataLakeServiceClient()` because it can return `Status::Invalid`.

Reply:
Falling back to the secure protocol is not so bad and we ensure it's valid when constructing from URI.

https://github.com/apache/arrow/blob/512d2458ff777f15354ee2a88bde7723ec3d5035/cpp/src/arrow/filesystem/azurefs.cc#L139-L145

Reply:
Passing an unsupported scheme should be a user error, since filesystem configuration is typically chosen by the user. It doesn't really help the user to allow incorrect values.

Reply:
> Should we validate the value explicitly? If we should validate the value, we'll do it in `MakeBlobServiceClient()` and `MakeDataLakeServiceClient()` because it can return `Status::Invalid`.

As long as the value is validated somewhere, I'm fine :-)

Reply:
OK. I'll do it as a separated task: GH-42134

Because this is not related to the Valgrind failure.